torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1979,6 +1983,906 @@ def clip_grad_norm_(
+        # Assume the standard case of passing `model.parameters()` to the
+        # optimizer if `optim_input` is not specified
+        if optim_input is None:
+            return list(model.parameters())
+        try:
+            params = list(optim_input)
+        except TypeError:
+            raise TypeError(
+                "Optimizer input should be an iterable of Tensors or dicts, "
+                f"but got {optim_input}"
+            )
+        if len(params) == 0:
+            raise ValueError("Optimizer input should not be empty")
+
+        # Check if the optimizer input represents tensors or parameter groups
+        all_tensors = True
+        all_dicts = True
+        for param in params:
+            all_tensors &= isinstance(param, torch.Tensor)
+            all_dicts &= isinstance(param, dict)
+        if not all_tensors and not all_dicts:
+            raise TypeError(
+                "Optimizer input should be an iterable of Tensors or dicts"
+            )
+        if all_tensors:
+            return params  # type: ignore[return-value]
+        assert all_dicts
+        flat_param_id_to_param = []
+        for param_group in params:
+            has_params_key = "params" in param_group  # type: ignore[operator]
+            assert has_params_key, \
+                "A parameter group should map \"params\" to a list of the " \
+                "parameters in the group"
+            for param in param_group["params"]:  # type: ignore[index]
+                # Implicitly map `flat_param_id` (current length of the list)
+                # to `param`
+                flat_param_id_to_param.append(param)
+        return flat_param_id_to_param  # type: ignore[return-value]
+
+    @staticmethod
+    def _preprocess_optim_state_dict(
+        optim_state_dict: Dict[str, Any],
+    ) -> Dict[str, Any]:
+        """
+        Pre-processes the rank's optimizer state dict by converting any
+        "step" state from a zero-dimension tensor to a float to support
+        backward compatibility with optimizer state dicts saved before version
+        1.12.
+
+        Args:
+            optim_state_dict (Dict[str, Any]): Optimizer state dict returned
+                from ``optim.state_dict()``, where ``optim`` is a rank's
+                optimizer.
+
+        Returns:
+            optim_state_dict (Dict[str, Any]): Pre-processed optimizer state
+                dict.
+
+        Raises:
+            ValueError: If there are any zero-dimension tensor states other
+                than "step" since we do not currently support them.
+        """
+        # Convert Adam's "step" state from a zero-dimension tensor to float if
+        # needed to support BC with dicts from before version 1.12
+        for param_state in optim_state_dict["state"].values():
+            if "step" in param_state:
+                step = param_state["step"]
+                if _is_zero_dim_tensor(step):
+                    param_state["step"] = step.item()
```

Comment:
Is this required? Optimizers in PyTorch have tensor based step and Pyper optimizer return a tensor based step too.